internal/policy/verify.go

```diff
@@ -0,0 +1,140 @@
+package policy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/adityasaky/gittuf/internal/gitinterface"
+	"github.com/adityasaky/gittuf/internal/rsl"
+	"github.com/adityasaky/gittuf/internal/tuf"
+	"github.com/go-git/go-git/v5"
+)
+
+var (
+	ErrUnauthorizedSignature = errors.New("unauthorized signature")
+)
+
+func VerifyRef(ctx context.Context, repo *git.Repository, target string) error {
+	// 1. Trace RSL back to the start
+	firstEntryTmp, err := rsl.GetFirstEntry(repo)
+	if err != nil {
+		return err
+	}
+	firstEntry := firstEntryTmp.(*rsl.Entry)
+
+	// 2. Find latest entry for target
+	latestEntryTmp, err := rsl.GetLatestEntryForRef(repo, target)
+	if err != nil {
+		return err
+	}
+	latestEntry := latestEntryTmp.(*rsl.Entry)
+
+	// 3. Do a relative verify from start entry to the latest entry
+	return VerifyRelativeForRef(ctx, repo, firstEntry, latestEntry, target)
+}
+
+func VerifyRelativeForRef(ctx context.Context, repo *git.Repository, firstEntry *rsl.Entry, lastEntry *rsl.Entry, target string) error {
+	entryStack := []*rsl.Entry{lastEntry}
+
+	var currentPolicy *State
+	// 1. Identify policy applicable at firstEntry
+	policyEntry, err := rsl.GetLatestEntryForRefBefore(repo, PolicyRef, firstEntry.ID)
+	if err != nil {
+		return err
+	}
+	state, err := LoadStateForEntry(ctx, repo, policyEntry)
+	if err != nil {
+		return err
+	}
+	currentPolicy = state
+
+	// 2. Enumerate RSL entries between firstEntry and lastEntry, ignoring irrelevant ones
+	iteratorEntry := lastEntry
+	for {
+		if iteratorEntry.GetID() == firstEntry.ID {
+			break
+		}
+
+		parentEntryTmp, err := rsl.GetParentForEntry(repo, iteratorEntry)
+		if err != nil {
+			return err
+		}
+		parentEntry := parentEntryTmp.(*rsl.Entry) // TODO: handle annotations
+
+		if parentEntry.RefName == target || parentEntry.RefName == PolicyRef {
+			entryStack = append(entryStack, parentEntry)
+		}
+
+		iteratorEntry = parentEntry
+	}
+
+	// entryStack has a list of RSL entries in reverse order
+	entryQueue := []*rsl.Entry{}
+	for j := len(entryStack) - 1; j >= 0; j-- {
+		// TODO: instead of reversing it, we can likely just protest this in
+		// reverse order but the policy states are the issue
```

Comment:
I don't fully understand this

Reply:
```
// We reverse the entries so that they are chronologically sorted. If we
// process them in the reverse order, we have to go past each entry
// anyway to find the last policy entry to use. It also makes the entry
// processing easier to reason about.
```

Updated the comment.